### ISSUE TITLE
Blockquote handling in html-dictionary. (Sametypesequence=h)

### ISSUE THREAD

**Issue by Markismus** [koreader/koreader#5859]
* KOReader version: 2020.01
* Device: linux/kobo Aura h2o

#### Issue
The indent on blockquote's is rather fierce. It's making a dictionairy such as Oxford Dictionary 2nd Edition non-readable on e-readers. Enlarging the koreader window to full screen shows that it is rendered nicely on a larger scale. 
What css controls the dictionairy display? Can we set the indent for blockquote as a percentage of the display width?

I've included some screen shots. Due to resizing it might not be directly apparent that the indent for the blockquote is of equal size in all screenshots.
Full screen:
![Screenshot from 2020-02-13 10-06-55](https://user-images.githubusercontent.com/5269101/74419637-868aa100-4e4a-11ea-8824-9e76c1eb0a1f.png)
![Screenshot from 2020-02-13 10-07-06](https://user-images.githubusercontent.com/5269101/74419642-87233780-4e4a-11ea-9755-f22fcfdddcd1.png)

Normal window:
![Screenshot from 2020-02-13 10-07-33](https://user-images.githubusercontent.com/5269101/74419644-87bbce00-4e4a-11ea-9816-44d6656cb066.png)
![Screenshot from 2020-02-13 10-07-49](https://user-images.githubusercontent.com/5269101/74419647-87bbce00-4e4a-11ea-9c77-d1608ad07052.png)
![Screenshot from 2020-02-13 10-08-17](https://user-images.githubusercontent.com/5269101/74419651-88ecfb00-4e4a-11ea-8457-8b9d9c9c3cfb.png)

#### Steps to reproduce
I've created a small excerpt from the Oxford dictionary and uploaded it [here.](https://markismus.stackstorage.com/s/Y6J6Xic3FVgWQHa) Lookup 'a'.

**Comment by Frenzie** [comment]
There's a CSS snippet in the dictionary or rich text widget file iirc. I'll check in a bit.

**Comment by Frenzie** [comment]
@Markismus It's not so much the default styles (something like margin-left & right `1em`) but the content of your dictionary:

```
<blockquote><blockquote><blockquote><blockquote><b><c c=\"indigo\">4.</c></b> Appended to lines ‘in burlesque poetry, to lengthen out a syllable, without adding to the sense.’ J. Not originally burlesque, but probably originating in the necessary retention of the <abr>ME.</abr> final <i>-e</i> where wanted for measure, the origin of which being forgotten, it was treated as an addition of <i>ă</i>. Thus <abr>ME.</abr> <i>sonne</i>—<i>yronne</i>, would be treated as <i>sun a!</i>—<i>run a!</i></blockquote></blockquote></blockquote></blockquote>
```

**Comment by poire-z** [comment]
No pb tweaking that blockquote in there, but dunno if we should hardcode dedicated workarounds in this main CSS snippet, like `blockquote blockquote { margin: 0; }`.
Is that "Oxford Dictionary 2nd Edition" version their shipped/published version? Or has it been converted to stardict from some other source?

Also note that one can do its own tweaking of each of its HTML dict with some dedicated .css or .lua file to be put alongside the .ifo. Links to the relevant issues in https://github.com/koreader/koreader/wiki/Dictionary-support _dictionaries can be tweaked with a custom CSS file_ - one can also massge the HTML with some Lua code.

As for the default for MuPDF, blockquote indeed does get a large 40px for margin-left/right. I commented about some of the defaults in the CSS for the FootnoteWidget:
https://github.com/koreader/koreader/blob/3e00e686b6fe76709fd6f0ac1696eebbac1e9fb5/frontend/ui/widget/footnotewidget.lua#L57-L91
(I guess I should fix that blockquote in there, as we have tweaked it in epub.css since then.)

**Comment by Markismus** [comment]
Yes, I know that the nested blockquotes mess things up. 

[Here are some default values for blockquotes](https://www.w3schools.com/tags/tag_blockquote.asp):
```sh
blockquote {
  display: block;
  margin-top: 1em;
  margin-bottom: 1em;
  margin-left: 40px;
  margin-right: 40px;
}
```
So the 40px definitely mess things up! Your .5em works nicely for this dictionary. However, for less densely nested dictionaries margin-left 1em works, too.


I've tested it a bit and it might be worth looking at a rather small right-margin. As dictionaries don't use actual quotes, but use blockquotes as a organizing tag, the space in the right doesn't increase legibility, but does cost space.

Both at 1em:
![Screenshot from 2020-02-13 11-47-45](https://user-images.githubusercontent.com/5269101/74427021-ade76b00-4e56-11ea-824c-56da825412b4.png)
margin-left:1em; margin-right:0em;
![Screenshot from 2020-02-13 11-48-21](https://user-images.githubusercontent.com/5269101/74427192-e424ea80-4e56-11ea-84cc-e08043d2c69e.png)

 


Goldendict runs into the same trouble:
![Screenshot from 2020-02-13 11-27-37](https://user-images.githubusercontent.com/5269101/74425371-dfab0280-4e53-11ea-853d-595a5bb88c45.png)

And linguae seems to strip all blockquotes as a solution:
![Screenshot from 2020-02-13 11-28-20](https://user-images.githubusercontent.com/5269101/74425428-f6e9f000-4e53-11ea-8c41-372cc23ffed1.png)

**Comment by Markismus** [comment]
> Is that "Oxford Dictionary 2nd Edition" version their shipped/published version? Or has it been converted to stardict from some other source?

It's the one from the [stardict tarball in the subdirectory _bigdics_](http://download.huzheng.org/bigdict/). It's been converted using stardict-bin2text and the resulting xml was combined into one dictionary, that one was snipped into less entries for easy testing and converted back using stardict-text2bin.

**Comment by Frenzie** [comment]
Browser default styles aren't terribly relevant, even if they're likely similar. See https://gitlab.com/koreader/mupdf/-/blob/9e90b9f90a946afb4634ea9e19d9a31ea7d58754/source/html/html-layout.c#L17-108 for ours. (They're mostly quite similar to your average browser styles of course.)

> but dunno if we should hardcode dedicated workarounds in this main CSS snippet, like blockquote blockquote { margin: 0; }.

Definitely not, that's a perfectly legit, unproblematic normal use case.

I'd just change it to `margin-left: 1em; margin-right: 0` and be done with it, possibly something smaller. The same applies to `dd`. If someone randomly nests fifty blockquotes for no apparent reason that's their problem. Perhaps we could try to be somewhat clever about it by using something like:

```css
blockquote blockquote {margin-left:.5em}
blockquote blockquote blockquote {margin-left:.3em}

**Comment by Markismus** [comment]
`"blockquote{display:block;margin:1em 40px}"`
Yes, the left and right margins were indeed 40px! Explains the shift.

Looking at the custom css, do I understand it correctly that I can only modify the style by using the lua-function html:gsub? 

So to alter the blockquote margin XMLLittre.lua would become:
```lua
-- XMLLittre HTML tweaks

return function(html)
--    f = io.open("/tmp/indict.html", "w")
--    f:write(html)
--    f:close()
    html = html:gsub('<blockquote>', '<blockquote style="margin-left:0.5em;margin-right:0;">')

--    f = io.open("/tmp/outdict.html", "w")
--    f:write(html)
--    f:close()
    return html
end
```

Or can I actually also just write css?

**Comment by poire-z** [comment]
You should be able to just create a `Oxford.blah.blah.css` (same base name as `Oxford.blah.blah.ifo`) containing the CSS you want prepended to all results, e.g.:
```css
blockquote { margin: 0 0 0 1em }
blockquote blockquote {margin-left:.5em}
blockquote blockquote blockquote {margin-left:.3em}
```
(The Lua `return function...` is for when you have uglier/invalid HTML that you really want to clean.)

**Comment by Frenzie** [comment]
In this case there are four nested blockquotes for no reason though, so you'd probably want either `blockquote blockquote {margin:0}` or `gsub("4 blockquotes", "just 1 blockquote")`.

My suggestion with subtly decreasing the margin is really just for proper use of blockquotes.

**Comment by poire-z** [comment]
> Reading through all the links, I saw your interest in Littré. I just created a stardict version of Nouveau Littré 2011.

Thank you :) Looks like definitions are quite a bit shorter than the one from the XMLLittre.

**Comment by Markismus** [comment]
Nice that you can also make css directly.

Can't figure out from [PetitRobert2007_1.1.lua.txt](https://github.com/koreader/koreader/files/1632075/PetitRobert2007_1.1.lua.txt) how to enable color in mupdf. Is this still the way it's done?

**Comment by poire-z** [comment]
Yes, `local colored = Screen:isColorEnabled()` should return true if you have `Gear> Screen> [x] Color rendering` checked (and obviously false on your eInk device :)

**Comment by poire-z** [comment]
Might be worth trying tweaking that unproper `<c c="indigo\>` via CSS, before going the Lua way.
`c[c=indigo] { color: indigo; } /* or the hex code */`

**Comment by Markismus** [comment]
No luck. I've propered it with the [adapted lua-file](https://gist.github.com/6f5b72d896073d81103a410dd2d2f7bd) (already done that before I read your comment...), mainly:
```lua
return function(html)
    html = html:gsub('<c c=\"', '<font color=\"')
    html = html:gsub('</c>', '</font>')
    html = html:gsub('<c>', '<font>')
    local colored = Screen:isColorEnabled()
    local stack = collect(html)
    
    html = serialize(stack, colored)
--    logger.warn("SERIALIZED:", html)

    f = io.open("/tmp/outdict.html", "w")
    f:write(html)
    f:close()

    return html
```
, and the generated [output.html ](https://gist.github.com/41598e8c058a375ca5a05b0e5510ab6e)works in firefox (with added html and body tags). 

However, even after checking the setting for screen>color rendering twice, no luck!

**Comment by poire-z** [comment]
It's possible `<font color=>` is not supported by MuPDF. Try `<span style=color: >` (with the appropriate balanced quotes).

**Comment by Markismus** [comment]
Very nice!
![Screenshot from 2020-02-13 13-53-37](https://user-images.githubusercontent.com/5269101/74437367-4803df00-4e68-11ea-8d53-000e94dec3ff.png)

I used:
```lua
html = html:gsub('</c>', '</span>')
    html = html:gsub('<c>', '<span>')
    html = html:gsub('<c c=\"indigo\">', '<span style=\"color:#4b0082\">')
    html = html:gsub('<c c=\"darkmagenta\">', '<span style=\"color:#8b008b\">')
    html = html:gsub('<c c=\"gray\">', '<span style=\"color:#808080\">')
    html = html:gsub('<c c=\"darkslategray\">', '<span style=\"color:#2f4f4f\">')
    ```

**Comment by Frenzie** [comment]
The named colors aren't supported?

**Comment by Markismus** [comment]
Only using these substitutions:
```lua
   html = html:gsub('</c>', '</span>')
    html = html:gsub('<c>', '<span>')
    html = html:gsub('<c c=\"', '<span style=\"color:')
```

generates:


![Screenshot from 2020-02-13 14-24-34](https://user-images.githubusercontent.com/5269101/74439485-99ae6880-4e6c-11ea-8716-fc55dc1b45ec.png)

**Comment by poire-z** [comment]
Yep, the only named color values in MuPDF sources are in `source/svg/svg-color.c` `svg_predefined_colors`, which does not seem to be used outside of this svg parser.

**Comment by Markismus** [comment]
I've converted them [in my script to hexvalues.](https://github.com/Markismus/PocketBookDic/commit/4d36e254f001f8923118ae9149b06455f5a79ec4)

**Comment by Frenzie** [comment]
@poire-z That's the main source of my surprise given that they're the same thing. I'd have expected svg_parse_color to be called, or to be moved into something more generic like css_parse_color. Oh well, it's probably just not come up yet and for the moment I'm not interested enough to send a patch upstream. :-P

**Comment by Frenzie** [comment]
Oh hold on, they do have CSS color keywords separately, just an extremely limited subset.

https://github.com/ArtifexSoftware/mupdf/blob/341ad550e43c9cc9e66a1d1cea571384d71f4eb5/source/html/css-apply.c#L1073-L1112

(Possibly the CSS1 keywords only?)

**Comment by Markismus** [comment]
[The result is beautiful](https://markismus.stackstorage.com/s/Tr8b5PTZ7DhrHkJ) on my Kobo. The shades of grey make the abundance of text much more clear. I would have made a screenshot, but I failed horribly at it while the dictionary window is open. 
(Is this just me? I kept moving the window or setting the library at default or skimming or whatever. I tried both diagonally swipping and tapping opposing corners.)
